llvm/lib/Analysis/LoopAccessAnalysis.cpp

```diff
@@ -2449,13 +2449,20 @@ bool LoopAccessInfo::analyzeLoop(AAResults *AA, const LoopInfo *LI,
         continue;
 
       // If this is a load, save it. If this instruction can read from memory
-      // but is not a load, then we quit. Notice that we don't handle function
-      // calls that read or write.
+      // but is not a load, we only allow it if it's a call to a function with a
+      // vector mapping and no pointer arguments.
       if (I.mayReadFromMemory()) {
-        // If the function has an explicit vectorized counterpart, we can safely
-        // assume that it can be vectorized.
+        auto hasPointerArgs = [](CallBase *CB) {
+          return any_of(CB->args(), [](Value const *Arg) {
+            return Arg->getType()->isPointerTy();
+          });
+        };
+
+        // If the function has an explicit vectorized counterpart, and does not
+        // take output/input pointers, we can safely assume that it can be
+        // vectorized.
         if (Call && !Call->isNoBuiltin() && Call->getCalledFunction() &&
-            !VFDatabase::getMappings(*Call).empty())
+            !hasPointerArgs(Call) && !VFDatabase::getMappings(*Call).empty())
```

Comment:
One thing that is a bit surprising here is that functions without pointer arguments are marked as mayReadFromMemory. If they don't access memory, shouldn't they be marked as not reading from memory?

Reply:
Perhaps `errno`? Although I'd expect that to be write only.

It's also possible that we're back in a position of effectively preventing all functions that access memory but given we have this finer control is there harm in keeping it?

Reply:
It is a bit, this seems to allow vectorizing some library calls without `-fno-math-errno` https://godbolt.org/z/KM4Pv6Pfn (though it seems to depend on the call not returning `float`).

Reply:
Functions might read from other parts of memory like globals or TU-local data (think lookup tables for some math functions). Or a system clock for a function like `omp_get_wtime()`, where not marking it as reading from memory led to failed tests since it was hoisted out of a loop.

There's the `ArgMemOnly` attribute to indicate that the only memory accesses are through pointers passed as arguments.

Reply:
Just trying to clarify the expected behavior, as it currently seems inconsistent. E.g. we vectorize `ilogbf` with ArmPL, but not `sin` without  `-fno-math-errno`: https://godbolt.org/z/xsGn1x5hd

At least on Linux, `ilogbf` may set errno (https://www.man7.org/linux/man-pages/man3/ilogb.3.html), so we should not vectorize with  `-fno-math-errno`, unless choosing ArmPL implies `-fno-math-errno` implicitly? 

If that is indeed the desired behavior, `ilogbf` should be marked as `readnone` (and we should drop the code here).

If that's not possible to do easily, the current change should be an OK start, but it should have a comment clearly stating which correctness issue is ignored for now

Reply:
I think the inconsistency comes from where this check is, `sin` is marked as `memory(write)` so `mayReadFromMemory()` is false, so it does not hit this check. `ilogbf` does not have it's attributes inferred (likely as it's a less well used function), so it's (conservatively) assumed that it `mayReadFromMemory()`, and it hits this check allowing it to vectorize it.

Reply:
It seems like the intention with this check was that if you have a vector mapping -> vectorize the call (regardless of the function attributes). It's not clear why it's gated under `mayReadFromMemory()` (there does not seem to be an explanation in the original patch: https://reviews.llvm.org/D8095).

The pointer arguments issue is a separate issue (which did not occur earlier as outside ArmPL/SLEEF there's no mapping for functions with output pointers).

Reply:
It would still be good to know if there is any justification for vectorizing cases with custom veclibs for functions that may access errno

but as I said earlier, fixing this can be done as follow up

Reply:
> It would still be good to know if there is any justification for vectorizing cases with custom veclibs for functions that may access errno

`-fveclib` has to be intentionally set by the user, which means they probably know what they're doing, and would like to force the use of the vector library calls (so don't care about `errno`). So one possible path forward is to move this check from under `mayReadFromMemory()` and allow it to apply more generally. This may also require updating the documentation of `-fveclib` to state this is intentional.

Reply:
If this is intentional, this should definitely be documented.

Reply:
(and then `fveclib` in clang should probably imply `-fno-math-errno`, so we don't need special handling in LLVM for that but can rely on the attributes set for the functions)

Reply:
I've posted a short RFC on this here: https://discourse.llvm.org/t/rfc-should-fveclib-imply-fno-math-errno-for-all-targets/81384